### PR TITLE
Integration tests for Tracing

### LINKED ISSUE #5156: Add end to end tracing testing for libraries (Http and Amqp)
Add more concrete and end to end tests for testing tracing in Amqp and Http libraries using the plugin package.

### PR DESCRIPTION
**Pull request by samvaity**
Fixes #5156, #5132

### CODE REVIEW

File: sdk/tracing/azure-core-tracing-opencensus/src/main/java/com/azure/core/tracing/opencensus/OpenCensusTracer.java [hunk 1]
```diff
@@ -258,7 +256,11 @@ private void addSpanRequestAttributes(Span span, Context context, String spanNam
      * @return The component name contained in the context.
      */
     private static String parseComponentValue(String spanName) {
-        return spanName.substring(spanName.indexOf(".") + 1, spanName.lastIndexOf("."));
+        if (spanName.length() > 0 && spanName.indexOf(".") != -1 && spanName.lastIndexOf(".") != -1) {
```
Comment (joshfree) [inline sdk/tracing/azure-core-tracing-opencensus/src/main/java/com/azure/core/tracing/opencensus/OpenCensusTracer.java]: the jit is hopefully going to optimize away these redundant calls to indexOf; but just in case, could you refactor the code to use temporary ints and avoid the duplicate API calls?

Comment (joshfree) [inline sdk/tracing/azure-core-tracing-opencensus/src/main/java/com/azure/core/tracing/opencensus/OpenCensusTracer.java]: should we guard against a null string as well?

File: sdk/tracing/azure-core-tracing-opencensus/src/main/java/com/azure/core/tracing/opencensus/implementation/AmqpTraceUtil.java [hunk 2]
```diff
@@ -24,7 +24,7 @@ public static Status parseStatusMessage(String statusMessage, Throwable error) {
                 ? Status.UNKNOWN.withDescription(message)
                 : Status.UNKNOWN.withDescription(error.getClass().getSimpleName());
 
-        } else if (statusMessage.equalsIgnoreCase("success")) {
+        } else if (statusMessage != null && statusMessage.equalsIgnoreCase("success")) {
```
Comment (joshfree) [inline sdk/tracing/azure-core-tracing-opencensus/src/main/java/com/azure/core/tracing/opencensus/implementation/AmqpTraceUtil.java]: given the first if-block returns, could you move this to the next line down as a regular if-block instead to make it slightly more clear.  Similarly, do you need to handle the statusMessage==null case separately so it doesn't also flow down to the Status.UNKNOWN.withDescription below?

File: sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 3]
```diff
@@ -51,27 +183,92 @@ public void startSpanOverloadNullPointerException() {
...
+    @Test
+    public void endSpanNoSuccessErrorMessageTest() {
+        // Arrange
+        RecordEventsSpanImpl recordEventsSpan = (RecordEventsSpanImpl) tracer.getCurrentSpan();
+
+        // Act
+        openCensusTracer.end(null, null, tracingContext);
+
+        // Assert
+        Assert.assertEquals(recordEventsSpan.getStatus().getCanonicalCode().toString(), "UNKNOWN");
```
Comment (joshfree) [inline sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: I know this is test code, but consider moving the string literals into a constant

**Comment by joshfree** [comment]
please be sure to squash this commit :)

File: sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 4]
```diff
@@ -51,27 +183,92 @@ public void startSpanOverloadNullPointerException() {
...
+        RecordEventsSpanImpl existingSpan = (RecordEventsSpanImpl) tracer.spanBuilder("existing-span").startSpan();
+        Context traceContext = tracingContext.addData(SPAN_CONTEXT, existingSpan.getContext());
+        RecordEventsSpanImpl parentSpanImpl = (RecordEventsSpanImpl) parentSpan;
+
+        // Act
+        openCensusTracer.addLink(traceContext);
+
+        //Assert
+        // TODO: existing -> parentSpanImpl comes after after add Link, confirm this behavior?
+        Assert.assertEquals(parentSpanImpl.getPrev(), existingSpan);
```
Comment (samvaity) [inline sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: @lmolkova I do not remember this very clearly but I think we introduced something called as a `Link.fromSpanContext` to mimic otel's behavior of adding Link.

But the way it should have worked is adding a link after the parentSpan, correct?

File: sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 5]
```diff
@@ -36,13 +77,111 @@ public void startSpanNullPointerException() {
...
+    }
+
+    @Test
+    public void startSpanProcessKindProcess() {
+        // Arrange
+        tracer.spanBuilder(OPENCENSUS_SPAN_KEY).startScopedSpan();
+        SpanId parentSpanId = parentSpan.getContext().getSpanId();
+
+        // Act
+        final Context updatedContext = openCensusTracer.start(METHOD_NAME, tracingContext, ProcessKind.PROCESS);
```
Comment (lmolkova) [inline sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: This typically happens when there is no parent, could we test it works?

File: sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 6]
```diff
@@ -36,13 +77,111 @@ public void startSpanNullPointerException() {
...
+
+        // verify span attributes
+        Map<String, AttributeValue> attributeMap = recordEventsSpan.toSpanData().getAttributes().getAttributeMap();
+        Assert.assertEquals(attributeMap.get(COMPONENT), AttributeValue.stringAttributeValue(COMPONENT_VALUE));
+        Assert.assertEquals(attributeMap.get(MESSAGE_BUS_DESTINATION), AttributeValue.stringAttributeValue(ENTITY_PATH_VALUE));
+        Assert.assertEquals(attributeMap.get(PEER_ENDPOINT), AttributeValue.stringAttributeValue(HOSTNAME_VALUE));
+    }
+
+    @Test
+    public void startSpanProcessKindReceive() {
```
Comment (lmolkova) [inline sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: I recall we decided against instrumenting Receive call

File: sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 7]
```diff
@@ -51,27 +190,97 @@ public void startSpanOverloadNullPointerException() {
...
+        Context traceContext = tracingContext.addData(SPAN_CONTEXT, existingSpan.getContext());
+        RecordEventsSpanImpl parentSpanImpl = (RecordEventsSpanImpl) parentSpan;
+
+        // Act
+        openCensusTracer.addLink(traceContext);
+
+        //Assert
+        // TODO: existing -> parentSpanImpl comes after after add Link, confirm this behavior?
+        Assert.assertEquals(parentSpanImpl.getPrev(), existingSpan);
+        Assert.assertEquals(parentSpanImpl.toSpanData().getChildSpanCount().intValue(), 1);
```
Comment (lmolkova) [inline sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: This is not related to links - not needed

File: sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java (hunk 4 above)
Comment (lmolkova) [reply sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: With opencensus, links are added one the span after span is created, with opentelemetry - before creation on builder.

What do you mean by 'after the parentSpan'?

File: sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java (hunk 6 above)
Comment (samvaity) [reply sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: Receive is the message span. 
I think with eventhubs we wanted to call it this way as "message" didn't macth any of th existing terms we used in the package.

File: sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java (hunk 4 above)
Comment (samvaity) [reply sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: So the way we have the link is, we have parentSpan followed by existing span.

Comment (samvaity) [reply sdk/tracing/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: Removed this and now checking if the two spans are related by the same TraceId

File: sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 8]
```diff
@@ -35,13 +79,119 @@ public void startSpanNullPointerException() {
...
+        final SpanId parentSpanId = parentSpan.getContext().getSpanId();
+
+        // Act
+        final Context updatedContext = openCensusTracer.start(METHOD_NAME, tracingContext);
+
+        // Assert
+        assertSpanWithExplicitParent(updatedContext, parentSpanId);
+        final RecordEventsSpanImpl recordEventsSpan =
+            (RecordEventsSpanImpl) updatedContext.getData(PARENT_SPAN_KEY).get();
+        Assert.assertNull(recordEventsSpan.getKind());
```
Comment (lmolkova) [inline sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: just curious, why check for null kind?

File: sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 9]
```diff
@@ -50,27 +200,99 @@ public void startSpanOverloadNullPointerException() {
...
+        final Context traceContext = tracingContext.addData(SPAN_CONTEXT_KEY, testSpan.getContext());
+        final RecordEventsSpanImpl parentSpanImpl = (RecordEventsSpanImpl) parentSpan;
+
         // Act
-        final Context updatedContext = openCensusTracer.start(METHOD_NAME, Context.NONE, ProcessKind.SEND);
+        openCensusTracer.addLink(traceContext);
+
+        //Assert
+        // verify both spans share the same link traceId
+        Assert.assertEquals(parentSpanImpl.toSpanData().getContext().getTraceId(),
```
Comment (lmolkova) [inline sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: this seems to be an incorrect validation. Links do not flow into trace-id.

So i guess what happens in this test:
setUp creates parent span
testSpan will get same TraceId - because it's a child
traceContext gets context from traceSpan
link is created from traceContext (which shares trace-id with parentSpan and testSpan
Then you check if parentSpan has the same traceId as link

What should be tested:
link should have random context - new one
you add it  to the span
and then compare links on the parent span to what you've added

File: sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java (hunk 8 above)
Comment (samvaity) [reply sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: Just to validate it isn't getting set anywhere in the workflow.

File: sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java (hunk 9 above)
Comment (samvaity) [reply sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: updated.

File: sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 10]
```diff
@@ -35,13 +79,119 @@ public void startSpanNullPointerException() {
...
+        Assert.assertEquals(attributeMap.get(PEER_ENDPOINT), AttributeValue.stringAttributeValue(HOSTNAME_VALUE));
+    }
+
+    @Test
+    public void startSpanProcessKindReceive() {
+        // Arrange
+        final SpanId parentSpanId = parentSpan.getContext().getSpanId();
+
+        // Act
+        final Context updatedContext = openCensusTracer.start(METHOD_NAME, tracingContext, ProcessKind.RECEIVE);
```
[the remark this answers is not in the document]
Comment (samvaity) [reply sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]: updated it to `message`.

File: sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java [hunk 11]
```diff
@@ -35,13 +80,119 @@ public void startSpanNullPointerException() {
...
+        final Map<String, AttributeValue> attributeMap = recordEventsSpan.toSpanData().getAttributes()
+            .getAttributeMap();
+        Assert.assertEquals(attributeMap.get(COMPONENT), AttributeValue.stringAttributeValue(COMPONENT_VALUE));
+        Assert.assertEquals(attributeMap.get(MESSAGE_BUS_DESTINATION),
+            AttributeValue.stringAttributeValue(ENTITY_PATH_VALUE));
+        Assert.assertEquals(attributeMap.get(PEER_ENDPOINT), AttributeValue.stringAttributeValue(HOSTNAME_VALUE));
+    }
+
+    @Test
+    public void startSpanProcessKindReceive() {
```
Comment (lmolkova) [inline sdk/core/azure-core-tracing-opencensus/src/test/java/com/azure/core/tracing/opencensus/OpenCensusTracerTest.java]:
```suggestion
    public void startSpanProcessKindMessage() {
```
